### PR TITLE
MS Teams send-notification command bug

### PR DESCRIPTION
**Pull request by merit-maita**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://jira-dc.paloaltonetworks.com/browse/XSUP-32880

## Description
- Fixed an issue where ***send-notification*** command failed when a message contains a GUID. 

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py [hunk 1]
```diff
@@ -502,6 +504,14 @@ def process_mirror_or_unknown_message(message: str) -> dict:
     return create_adaptive_card(body)
 
 
+def is_teams_ask_message(msg: str) -> bool:
+    try:
+        message: dict = json.loads(msg)
+        return MS_TEAMS_ASK_MESSAGE_KEYS == message.keys()
```
Comment (dorschw) [inline Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py:510]:
```suggestion
        return set(message.keys()) == MS_TEAMS_ASK_MESSAGE_KEYS 
```

File: Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams_test.py [hunk 2]
```diff
@@ -550,7 +560,8 @@ def test_send_message_with_entitlement(mocker, requests_mock):
         'options': ['yes', 'no', 'maybe'],
         'entitlement': '4404dae8-2d45-46bd-85fa-64779c12abe8',
         'investigation_id': '72',
-        'task_id': '23'
+        'task_id': '23',
+        'form_type': 'predefined-options'
```
Comment (dorschw) [inline Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams_test.py:564]: why this change?

File: Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams_test.py [hunk 3]
```diff
@@ -2340,3 +2351,30 @@ def test_handle_teams_proxy_and_ssl(mocker, is_xsoar_8, expected_result):
 
     proxies, use_ssl = ms_teams.handle_teams_proxy_and_ssl()
     assert (proxies, use_ssl) == expected_result
+
+
+@pytest.mark.parametrize('message, result', [
+    ('{"message_text": "message", "options": ["option"], "entitlement": "id", "investigation_id": "inv_id", "task_id": "task", '
+     '"form_type": "form"}', True),
+    ('{"message_text": "message", "options": ["option"], "entitlement": "id", "investigation_id": "inv_id", "task_id": "task",'
+     ' "form_type": "form", "extra_key": "extra"}', False),
```
Comment (dorschw) [inline Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams_test.py:2360]:
```suggestion

DUMMY_ASK_MESSAGE = {"message_text": "message", "options": ["option"], "entitlement": "id", "investigation_id": "inv_id", "task_id": "task", "form_type": "form"}

@pytest.mark.parametrize('message, result', [
    (json.dumps(DUMMY_ASK_MESSAGE), True),
    (json.dumps(DUMMY_ASK_MESSAGE | {"extra_key": "extra"}), False),
```

File: Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams_test.py [hunk 4]
```diff
@@ -2340,3 +2351,30 @@ def test_handle_teams_proxy_and_ssl(mocker, is_xsoar_8, expected_result):
...
+    ("non json message", False)
+])
+def test_is_teams_ask_message(message, result):
+    """
+    Given:
+        - A message in json format that has the exact same keys as ms teams ask card keys.
+        - A message in json format that has all ms teams ask card keys and an extra key.
+        - A message that's not json format.
```
Comment (dorschw) [inline Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams_test.py:2368]:
```suggestion
        - input message string
```

File: Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams_test.py [hunk 5]
```diff
@@ -2340,3 +2351,30 @@ def test_handle_teams_proxy_and_ssl(mocker, is_xsoar_8, expected_result):
...
+    When:
+        - Running is_teams_ask_message.
+    Then:
+        - Assert that the message is a legit ms teams ask message (from ms teams ask script flow).
+        - Assert that the message is not a message from ms teams ask script.
+        - Assert that the message is not a message from ms teams ask script.
+    If the test fails, please update the first message param in the test to have the same keys as MS_TEAMS_ASK_MESSAGE_KEYS
```
Comment (dorschw) [inline Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams_test.py:2375]:
```suggestion
        - Assert only ask_teams messages return True
    If the test fails, please update the first message param in the test to have the same keys as MS_TEAMS_ASK_MESSAGE_KEYS
```

File: Packs/MicrosoftTeams/Scripts/MicrosoftTeamsAsk/MicrosoftTeamsAsk_test.py [hunk 6]
```diff
@@ -3,9 +3,14 @@
 import demistomock as demisto
 import json
 import pytest
+from Packs.MicrosoftTeams.Integrations.MicrosoftTeams.MicrosoftTeams import MS_TEAMS_ASK_MESSAGE_KEYS
```
Comment (dorschw) [inline Packs/MicrosoftTeams/Scripts/MicrosoftTeamsAsk/MicrosoftTeamsAsk_test.py:6]: I don't think we can do that :( Which is why we discussed two unit tests, here and in Teams, checking about the same thing.

File: Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams_test.py (hunk 2 above)
Comment (merit-maita) [reply Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams_test.py:564]: it was missing this required key in the ask message, (this key was added a couple of months ago, so i guess the test was not updated), but when i added the validation to the ask keys, it failed

File: Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py [hunk 7]
```diff
@@ -502,6 +504,14 @@ def process_mirror_or_unknown_message(message: str) -> dict:
     return create_adaptive_card(body)
 
 
+def is_teams_ask_message(msg: str) -> bool:
+    try:
+        message: dict = json.loads(msg)
+        return set(message.keys()) == MS_TEAMS_ASK_MESSAGE_KEYS
```
Comment (dorschw) [inline Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py:510]:
```suggestion
        return message.keys() == MS_TEAMS_ASK_MESSAGE_KEYS 
```
my bad 😅 it's actually comparable as a set

File: Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py [hunk 8]
```diff
@@ -103,6 +103,8 @@ class FormType(Enum):  # Used for 'send-message', and by the MicrosoftTeamsAsk s
 TOKEN_EXPIRED_ERROR_CODES = {50173, 700082, }  # See: https://login.microsoftonline.com/error?code=
 REGEX_SEARCH_ERROR_DESC = r"^[^:]*:\s(?P<desc>.*?\.)"
 
+MS_TEAMS_ASK_MESSAGE_KEYS = {'message_text', 'options', 'entitlement', 'investigation_id', 'task_id', 'form_type'}
```
Comment (dorschw) [inline Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py:106]:
```suggestion
MS_TEAMS_ASK_MESSAGE_KEYS = {'message_text', 'options', 'entitlement', 'investigation_id', 'task_id', 'form_type'} # must be synced with ones in TeamsAsk
```

File: Packs/MicrosoftTeams/Scripts/MicrosoftTeamsAsk/MicrosoftTeamsAsk_test.py [hunk 9]
```diff
@@ -4,8 +4,14 @@
 import json
 import pytest
 
+MS_TEAMS_ASK_MESSAGE_KEYS = {'message_text', 'options', 'entitlement', 'investigation_id', 'task_id', 'form_type'}
```
Comment (dorschw) [inline Packs/MicrosoftTeams/Scripts/MicrosoftTeamsAsk/MicrosoftTeamsAsk_test.py:7]:
```suggestion
MS_TEAMS_ASK_MESSAGE_KEYS = {'message_text', 'options', 'entitlement', 'investigation_id', 'task_id', 'form_type'}  # must be synced with ones in MicrosoftTeams.py
```

**Review by dorschw — APPROVED** [review]
